sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionedTablePerfStatsSuite.scala

```diff
@@ -352,4 +353,28 @@ class PartitionedTablePerfStatsSuite
       }
     }
   }
+
+  test("SPARK-18700: add lock for each table's realation in cache") {
+    withTable("test") {
+      withTempDir { dir =>
+        HiveCatalogMetrics.reset()
+        setupPartitionedHiveTable("test", dir)
+        // select the table in multi-threads
+        val executorPool = Executors.newFixedThreadPool(10)
+        (1 to 10).map(threadId => {
+          val runnable = new Runnable {
+            override def run(): Unit = {
+              spark.sql("select * from test where partCol1 = 999").count()
+            }
+          }
+          executorPool.execute(runnable)
+          None
+        })
+        executorPool.shutdown()
+        executorPool.awaitTermination(30, TimeUnit.SECONDS)
+        // check the cache hit, the cache only load once
+        assert(HiveCatalogMetrics.METRIC_DATASOUCE_TABLE_CACHE_HITS.getCount() == 9)
```

Comment:
It may failed sometimes without lock, but when I add `Thread.sleep(1000)` before `cachedDataSourceTables.put(tableIdentifier, created)` in HiveMetastoreCatalog.scala +265 to make the build table relation slow.It will failed every time. How can I do this hook in UT? Or how to make the cache build operation slow without really make a big table?  : )